# GitHub



# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:**  Eduardo Oñate y Sebastián Versluys

**Fecha límite de entrega 📆:** Martes 4 de Abril.

**Tiempo estimado de dedicación:** 




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 100 palabras** . 🙏
` ` 

**Pregunta 1 (0.2 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones según lo visto en clases. Debe citar la slide donde cree que está la respuesta.**




` ` 
(Basándonos en la slide 16 de la introducción (Challenges of Language):

**Discreteness:** en nuestro idioma no se puede comprender el significado de una palabra según el mero orden de sus letras. Por tal razón un modelo simple que solo tenga en cuenta el significante del lenguaje humano sin su significado carecerá de aplicaciones.

**Compositionality:** existen infinitas combinaciones posibles de palabras para crear oraciones de significados o interpretaciones muy variadas y concretas (o no) dependiendo del contexto. Por ende un análisis más simple y secuencial (por ejemplo, basado en bigramas o trigramas) puede ser una opción muy pobre para ciertas aplicaciones.
` ` 


**Pregunta 2 (0.4 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.


` ` 
Por una parte el empirismo requiere la definición de características medibles con base lingüística acerca de un texto para ser posteriormente tratadas estadísticamente. Si bien es un método que se apoya en recursos computacionales para su cálculo y tratamiento, es un proceso en esencia *human made*.

Por otro lado, el Deep Learning trata de encontrar representaciones vectoriales de un texto <<"a fuerza bruta">> propagándolo por sus capas y minimizando su función de *loss*, encontrando en él relaciones no definidas previamente por humanos. Gracias a su increíble capacidad de cómputo estas redes superan en desempeño a los métodos empíricos actualmente.
` ` 

 **Pregunta 3 (0.4 puntos) Según las primeras clases, ¿Qué método clásico nos permite rankear las similitudes existentes entre documentos?, ¿Cómo son las representaciones que genera y problemas que podrían experimentar estas soluciones simples?** 

` ` 
Para la **bolsa de palabras** se identifica el vocabulario $V$ de un corpus. Luego, todos los documentos $d_i$ a analizar pueden ser representados como vectores de la misma dimensionalidad $|V|$, donde cada coordenada corresponde a un término $t_j$ del vocabulario, cuantificando su relevancia en cada documento. Luego, los vectores pueden ser rankeados según métricas como la distancia euclidiana o su similitud coseno.

La desventaja del método anterior es que implica la pérdida del **orden** de las palabras y de la estructura lingüística de un documento, careciendo de parte importante del contexto de habla y no siendo suficiente en algunas aplicaciones.
` ` 

**Pregunta 4 (0.4 puntos) Usted se encuentra realizando un modelo de clasificación de sentimientos con texto, su jefe le señala que debe eliminar las palabras mas comunes para obtener una mejor clasificación. ¿Qué palabras le señala que elimine su jefe?, ¿es acaso esto una buena idea?.** Puede utilizar ChatGPT (debe indicarlo) para generar la respuesta y luego debe indicar si la respuesta entregada por ChatGPT es correcta o no. Mencione por qué según lo visto en clases.

` ` 
La eliminación de stopwords podría reducir el ruido de los datos. Sin embargo, dependerá mucho de las palabras definidas como tal. En particular, algunas palabras podrían ser consideradas stopwords por su alta frecuencia de aparición (p. ej. "no", "ni", "tampoco", etc.), a pesar de tener un contenido semántico adversativo, interpretándose el documento de manera opuesta.

Por lo anterior, estamos de acuerdo con ChatGPT al proponer realizar experimentos con y sin eliminación de stopwords. Además, recomendaríamos variar la definición de stopwords con distintos umbrales y considerando otros aspectos lingüísticos como los mencionados anteriormente para adaptarlo a la aplicación requerida.
` ` 

**Pregunta 5 (0.6 puntos): Acorde al paper [A Vector Space Model for automatic indexing](https://dl.acm.org/doi/pdf/10.1145/361219.361220) un documento, $D_i$, puede ser definido formalmente como una tupla de términos, $(d_{i1}, d_{i2}, ..., d_{in})$, donde $d_{ij}$ representa el peso del j-esimo término. En clase vieron algunas formas medir los pesos de estos términos. Mencione cuales fueron y sus ventajas y desventajas.** 

` ` 
Usando **álgebra booleana** se tiene que para un corpus $C$ con vocabulario $V$, un término $t_j \in V$ y un documento $D_i \in C$:

$$
d_{i,j} = 
\begin{cases}
1, \text{ si } t_j \in D_i\\
0, \text{ si } t_j \notin D_i
\end{cases}.
$$

Este método es extremadamente simple de computar, aunque no discrimina la relevancia relativa entre distintos términos en un documento.

Para remediarlo se usan dos factores:


* La **frecuencia de términos** $(tf)$ normaliza y compara la frecuencia del término $t_j$ en el documento $D_i$. Mientras mayor sea, mayor relevancia.

* La **frecuencia invertida de documentos** $(idf)$ caracteriza el porcentaje de documentos en un corpus donde aparece un término $t_j$. Mientras menor sea esta cantidad, mayor es el $idf$ y mayor relevancia.

Luego, ambas métricas se multiplican para calcular el peso:

$$
d_{i,j} = tf_j × idf_j.
$$

A pesar de sus ventajas, esta última representacion de los documentos sigue siendo una variación de la **bolsa de palabras**, donde gran parte del contexto lingüístico de un texto es eliminado, no siendo suficiente para algunas aplicaciones.


` ` 

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Tokens*.
- El **orden de los resultados** pueden variar dependiendo de su máquina, pero los valores de los resultados son los mismos.



**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [2]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [3]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [4]:
import codecs
import re
import numpy as np
import pandas as pd

En caso de desarrollar la tarea desde colab, utilizar el siguiente código para cargar los archivos desde el drive:

` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene la letra de una canción del marcianeke 👽. Utilice este texto para realizar su primera tokenización y ver qué tan bien funciona su función. 

Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [5]:
text = codecs.open('marcianeke.txt', 'r', 'UTF-8').read()
print(text)

Brr
Marcianeke
Vamo' a estar con Pailita
Dimelo má
Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar (brr)
Tussi, keta quiere' mezclar
Dimelo má

Ando en busca de una criminal (ah, ah)
Esa que el gatillo le gusta jalar (rata-ta)
Que le guste flotar y fumar
Tussi, keta pura quiere' mezclar
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má
Di-dimelo má

Esperame que ahora entro yo
Y lo que pide yo lo traje
No visto de traje
Puro corte calle, no de maquillaje
Pronto coronamos y nos vamo' de viaje
Tanto hit que hago que lo' culo bajen
Ella se va de shopping
Sale positivo si se hace el doping
Baila twerk con un poco de popping
Los fardos en el botín

Si quieren letra llamen pa' mi booking
Generando, sigo en la mía lowkey
Cooking en el estudio con tu woman
Tanto whisky, pisco que hasta lo' vecinos toman
Si se tiran pa' aca puede que la arena coman
Ja, en el chanteo titulado sin diploma
Di-dimelo má
Di-di

Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [6]:
def get_tokens(text):
  ### Inicio del código ###"

  recopilador_palabras = []

  symbols = '!"#$%&/()=?¡¿\/´+}{-_[¨~^`|¬°*].,;:'
  symbol_2 = "'"
  saltos_linea = '\n\r'
  saltos_linea_2 = '\r'
  espacio_entre_parrafos = ''
  separador_palabras = ' '
  
  inicio_palabra = 0

  for fin_palabra in range(len(text)):

    espacio = text[fin_palabra] == separador_palabras
    simbolo = (text[fin_palabra] in symbols) or text[fin_palabra] == symbol_2
    salto_de_linea = text[fin_palabra] in saltos_linea
    
    if espacio or simbolo or salto_de_linea:

      nueva_palabra =  text[inicio_palabra:fin_palabra]

      # Si es un espacio o un simbolo guardamos la palabra que antecede
      if nueva_palabra != '':
        recopilador_palabras.append(nueva_palabra)
      
      # Si era un simbolo lo agregamos
      if simbolo:
        recopilador_palabras.append(text[fin_palabra])
      
      inicio_palabra = fin_palabra + 1

  # Al final del texto hay que corroborar que exista una palabra
  if (not simbolo) and (not salto_de_linea):
    palabra_a_agregar = text[inicio_palabra:fin_palabra+1]
    recopilador_palabras.append(palabra_a_agregar)

  return recopilador_palabras

  ### Fin del código ###

In [26]:
tokens = get_tokens(text)

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

Los supuestos que se tomaron para la tokenización, fue que todos los símbolos que se tomaban en consideración eran: ['!"#$%&/()=?¡¿\/´+}{-_[¨~^`|¬°*].,;:'], además de lso saltos de linea '\n' y '\r'. Se tuvo que crear una variable aparte que considerara la comilla única, es decir "'".

Se considera que todas las palabras que se le pasó a la función existen o tienen un significado linguístico. Además que no existan números entre medio de las palabras.


In [28]:
from nltk.tokenize import wordpunct_tokenize 

nltk_tokens = wordpunct_tokenize(text)
nltk_tokens == tokens

True

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [40]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los tokens de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [41]:
def get_vocab(dataset):
  ### Aquí inicia tu código ###

  lista_palabras = []

  for frases in dataset:
    tokens_por_frase = get_tokens(frases)
    
    for token in tokens_por_frase:

      if token not in lista_palabras:
        lista_palabras.append(token)

  return lista_palabras
  ### Aquí termina tu código ###

***Test:***

In [42]:
vocab = get_vocab(dataset)
vocab

['I',
 'like',
 'human',
 'languages',
 'programming',
 'Spanish',
 'is',
 'my',
 'favorite',
 'language']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

    Explique sus reglas aquí:

    Stopwords: Se eliminarán las palabras con menos de 2 letras, ya que se considerarán como pronombres o artículos.

    Stemming: Si existe el uso de una palabra tanto en singular como en plural, solo se tomará el caso singular. Para ello se verificarán aquellas palabras terminadas en s.


In [46]:
def pre_processing(vocab):
  ### Aquí inicia tu código ###
  for palabra in vocab:

    if len(palabra) <= 2:
      vocab.remove(palabra)

    elif palabra[-1] == 's':
      if palabra[:-1] in vocab:
        vocab.remove(palabra)
  
  return vocab

  ### Aquí termina tu código ###
    

In [47]:
vocab = pre_processing(vocab)
vocab

['like', 'human', 'programming', 'Spanish', 'favorite', 'language']


**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
 


Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

**disclaimer: El orden de los resultados pueden variar**

In [16]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [17]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ###

  resultado = pd.DataFrame(dtype = int)

  # Se recorren todos los datasets
  for frases in dataset:

    # Diccionario servira para llevar la cuenta de las palabras
    # palabras servira para saber que palabras se han usado
    diccionario = {}
    key_words = []

    # Se obtienen los tokens de cada palabra
    palabras = get_tokens(frases)
    for token in palabras:
      # Si la palabra no fue utilizada, se crea el contador en el diccionario
      if token not in key_words:
        diccionario[token] = [1]
        key_words.append(token)

      # Si la palabra ya se usó, se agrega 1 al contador
      else:
        diccionario[token][0] += 1

    resultado_temp = pd.DataFrame(diccionario,dtype = int)
    resultado = pd.concat([resultado,resultado_temp]).fillna(0).astype(int)

  lista_filas = pd.Index([f'd{di}' for di in range(len(dataset))])
  resultado = resultado.set_index(lista_filas, drop=True)

  return resultado
  
  ### Aquí termina tu código ###

***Test:***

In [18]:
dataset_bow = bag_of_words(dataset)
dataset_bow

El  perro  se  come  la  comida  y  después  duerme  despierta  empieza  \
d0   1      1   2     1   1       1  1        1       1          0        0   
d1   1      1   1     0   0       0  1        1       0          1        1   
d2   1      1   1     1   1       1  1        1       0          0        0   
d3   1      0   2     1   1       1  1        1       1          0        0   
d4   1      0   1     0   0       0  1        1       0          1        1   
d5   1      0   1     1   1       1  1        1       0          0        0   

    a  ladrar  ladra  gato  maullar  maulla  
d0  0       0      0     0        0       0  
d1  1       1      0     0        0       0  
d2  0       0      1     0        0       0  
d3  0       0      0     1        0       0  
d4  1       0      0     1        1       0  
d5  0       0      0     1        0       1

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector. 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [19]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    for fila in dataset_bow.index:
      df_row = dataset_bow.loc[fila]
      row_max = max(df_row)
      dataset_bow.loc[fila] = df_row/row_max
    return dataset_bow
    ### Aquí termina tu código ###

***Test:***

In [20]:
tf = calc_tf(dataset_bow)
tf

El  perro   se  come   la  comida    y  después  duerme  despierta  \
d0  0.5    0.5  1.0   0.5  0.5     0.5  0.5      0.5     0.5        0.0   
d1  1.0    1.0  1.0   0.0  0.0     0.0  1.0      1.0     0.0        1.0   
d2  1.0    1.0  1.0   1.0  1.0     1.0  1.0      1.0     0.0        0.0   
d3  0.5    0.0  1.0   0.5  0.5     0.5  0.5      0.5     0.5        0.0   
d4  1.0    0.0  1.0   0.0  0.0     0.0  1.0      1.0     0.0        1.0   
d5  1.0    0.0  1.0   1.0  1.0     1.0  1.0      1.0     0.0        0.0   

    empieza    a  ladrar  ladra  gato  maullar  maulla  
d0      0.0  0.0     0.0    0.0   0.0      0.0     0.0  
d1      1.0  1.0     1.0    0.0   0.0      0.0     0.0  
d2      0.0  0.0     0.0    1.0   0.0      0.0     0.0  
d3      0.0  0.0     0.0    0.0   0.5      0.0     0.0  
d4      1.0  1.0     0.0    0.0   1.0      1.0     0.0  
d5      0.0  0.0     0.0    0.0   1.0      0.0     1.0

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [21]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###

    idf_dic = {}
    doc_num = len(dataset_bow.index)
    for columnas in dataset_bow:
      n_i = (dataset_bow[columnas] > 0)*1
      idf_dic[columnas] = np.log10(doc_num/n_i.sum())
    
    return idf_dic
    ### Aquí termina tu código ###

***Test:***

In [22]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'come': 0.17609125905568124,
 'la': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'y': 0.0,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'despierta': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'a': 0.47712125471966244,
 'ladrar': 0.7781512503836436,
 'ladra': 0.7781512503836436,
 'gato': 0.3010299956639812,
 'maullar': 0.7781512503836436,
 'maulla': 0.7781512503836436}

***Resultado esperado***: 

```python
{'El': 0.0, 
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}
```

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [23]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###

    return tf*idf

    ### Aquí termina tu código ### 

***Test.***

In [24]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

El     perro   se      come        la    comida    y  después    duerme  \
d0  0.0  0.150515  0.0  0.088046  0.088046  0.088046  0.0      0.0  0.238561   
d1  0.0  0.301030  0.0  0.000000  0.000000  0.000000  0.0      0.0  0.000000   
d2  0.0  0.301030  0.0  0.176091  0.176091  0.176091  0.0      0.0  0.000000   
d3  0.0  0.000000  0.0  0.088046  0.088046  0.088046  0.0      0.0  0.238561   
d4  0.0  0.000000  0.0  0.000000  0.000000  0.000000  0.0      0.0  0.000000   
d5  0.0  0.000000  0.0  0.176091  0.176091  0.176091  0.0      0.0  0.000000   

    despierta   empieza         a    ladrar     ladra      gato   maullar  \
d0   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
d1   0.477121  0.477121  0.477121  0.778151  0.000000  0.000000  0.000000   
d2   0.000000  0.000000  0.000000  0.000000  0.778151  0.000000  0.000000   
d3   0.000000  0.000000  0.000000  0.000000  0.000000  0.150515  0.000000   
d4   0.477121  0.477121  0.477121  0.000000  0.000000  0.301030  0.778151   
d5   0.000000  0.000000  0.000000  0.000000  0.000000  0.301030  0.000000   

      maulla  
d0  0.000000  
d1  0.000000  
d2  0.000000  
d3  0.000000  
d4  0.000000  
d5  0.778151

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [86]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###

  numerador_distancia_coseno = (v1*v2).sum()
  denominador_distancia_coseno = (np.sqrt((v1**2).sum())) * (np.sqrt((v2**2).sum()))
  
  return abs(round(1 - (numerador_distancia_coseno/denominador_distancia_coseno),5))

  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity

display(similarity_matrix)

array([[0.     , 0.87968, 0.67766, 0.22033, 1.     , 0.83672],
       [0.87968, 0.     , 0.91314, 1.     , 0.50479, 1.     ],
       [0.67766, 0.91314, 0.     , 0.83672, 1.     , 0.88212],
       [0.22033, 1.     , 0.83672, 0.     , 0.87968, 0.67766],
       [1.     , 0.50479, 1.     , 0.87968, 0.     , 0.91314],
       [0.83672, 1.     , 0.88212, 0.67766, 0.91314, 0.     ]])

Como sabemos en la similitud coseno, entre más grande sea el valor más similitud tendrán ambos vectores. En este caso existen 2 donde tienen la misma similitud, los cuales se muestran a continuación:

- d1 y d2, con una similitud de 0.91314
- d4 y d5, con una similitud de 0.91314

![gato](https://live.staticflickr.com/4652/38904147065_0b6c446945_b.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**